core/tests/mod.rs

```diff
@@ -0,0 +1,778 @@
+
+use iroha_data_model::prelude::*;
+use iroha_data_model::block::*;
+use iroha_genesis::{GenesisNetwork, RawGenesisBlock, GenesisNetworkTrait};
+use iroha_core::smartcontracts::Registrable;
+use rand::seq::IteratorRandom;
+use rand::SeedableRng;
+
+use iroha_core::sumeragi::main_loop::Sumeragi;
+
+use iroha_data_model::transaction::Accept;
+
+
+/// Get a standardised key-pair from the hard-coded literals.
+///
+/// # Panics
+/// Programmer error. Given keys must be in proper format.
+pub fn get_key_pair() -> KeyPair {
+    KeyPair::new(
+        PublicKey::from_str(
+            "ed01207233BFC89DCBD68C19FDE6CE6158225298EC1131B6A130D1AEB454C1AB5183C0",
+        )
+        .expect("Public key not in mulithash format"),
+        PrivateKey::from_hex(
+            Algorithm::Ed25519,
+            "9AC47ABF59B356E0BD7DCBBBB4DEC080E302156A48CA907E47CB6AEA1D32719E7233BFC89DCBD68C19FDE6CE6158225298EC1131B6A130D1AEB454C1AB5183C0".as_ref(),
+        ).expect("Private key not hex encoded")
+    ).expect("Key pair mismatch")
+}
+
+static GENESIS_KEY : Mutex<Option<KeyPair>> = Mutex::new(None);
+fn genesis_key() -> KeyPair {
+    unsafe {
+    let mut mutex = GENESIS_KEY.lock().unwrap();
+    if let Some(key) = mutex.as_ref() {
+        key.clone()
+    } else {
+        let key = KeyPair::generate().unwrap();
+        *mutex = Some(key.clone());
+        key
+    }
+    }
+}
+
+fn genesis_account(public_key: iroha_crypto::PublicKey) -> Account {
+    Account::new(AccountId::genesis(), [public_key]).build(&AccountId::genesis())
+}
+
+fn genesis_world(peers: Vec<PeerId>) -> World {
+    let account_public_key = genesis_key().public_key().clone();
+    let mut domain = Domain::new(DomainId::genesis()).build(&AccountId::genesis());
+
+    domain.accounts.insert(
+        <Account as Identifiable>::Id::genesis(),
+        genesis_account(account_public_key.clone()),
+    );
+    
+    World::with(vec![domain], peers,)
+}
+
+fn sleep_ms(t: u64) {
+    std::thread::sleep(std::time::Duration::from_millis(t));
+}
+
+/// Trait used to differentiate a test instance of `genesis`.
+pub trait TestGenesis: Sized {
+    /// Construct Iroha genesis network and optionally submit genesis
+    /// from the given peer.
+    fn test(cfg: Configuration, submit_genesis: bool) -> Option<Self>;
+}
+
+static mut GENESIS_CACHE : Mutex<Option<RawGenesisBlock>> = Mutex::new(None);
+
+impl TestGenesis for GenesisNetwork {
+    fn test(cfg: Configuration, submit_genesis: bool) -> Option<Self> {
+        // TODO: Fix this somehow. Probably we need to make `kagami` a library (#3253).
+        let manifest_dir = Path::new(env!("CARGO_MANIFEST_DIR"));
+        let content = std::fs::read_to_string(manifest_dir.join("../configs/peer/genesis.json"))
+            .expect("Failed to read data from configs/peer/genesis.json");
+        
+        let mut genesis: RawGenesisBlock = unsafe {
```

Comment:
I think code like this needs `SAFETY` comment. Probably you should guarantee that this function will never be called in parallel or somtehing like that

Reply:
I switched it to be lazy initialized.